File: templates/typescript-protocol-adapter/test/metrics.test.ts
import { Config, MetricService } from "@edgecommons/edgecommons";
import { describe, expect, it } from "vitest";

import { Health } from "../src/app";
import {
  COMMAND,
  CONNECTION,
  DeviceMetrics,
  HEALTH,
  HEALTH_MEASURES,
  Pair,
  familyDefs,
} from "../src/metrics";

// --- a recording MetricService: captures every emitted (metric, values) pair --------------------

class RecordingMetrics implements MetricService {
  readonly emissions: Array<{ name: string; values: Record<string, number> }> = [];
  defineMetric(): void {}
  isMetricDefined(): boolean {
    return true;
  }
  async emitMetric(name: string, values: Record<string, number>): Promise<void> {
    this.emissions.push({ name, values });
  }
  async emitMetricNow(name: string, values: Record<string, number>): Promise<void> {
    this.emissions.push({ name, values });
  }
  async flushMetrics(): Promise<void> {}
  async shutdown(): Promise<void> {}
  last(name: string): Record<string, number> {
    return [...this.emissions].reverse().find((e) => e.name === name)?.values ?? {};
  }
}

function testConfig(): Config {
  return Config.fromValue("com.example.MyAdapter", "thing-1", {
    metricEmission: { target: "log", namespace: "test" },
    component: { global: {}, instances: [{ id: "plc-1" }] },
  });
}

const SECTION_5 = new Set([
  "connectionState",
  "publishLatencyMs",
  "pollLatencyMs",
  "readErrors",
  "staleSignals",
  "reconnects",
]);

describe("southbound_health parity (SOUTHBOUND.md §5)", () => {
  it("emits EXACTLY the §5 measure set — no more, no less", () => {
    const health = familyDefs().find((f) => f.name === HEALTH);
    expect(health).toBeDefined();
    const emitted = new Set(health!.measures.map((m) => m.name));
    expect(emitted).toEqual(SECTION_5);
    // The advertised const must agree with what familyDefs emits.
    expect(new Set(HEALTH_MEASURES)).toEqual(SECTION_5);
  });
});

describe("the operational-family pattern", () => {
  it("names the families from the component and keeps only low-cardinality dimensions", () => {
    const defs = familyDefs();
    const names = defs.map((f) => f.name);
    expect(names).toContain(CONNECTION);
    expect(names).toContain(COMMAND);
    // Named from the component token — a fleet view separates adapters by name.
    expect(CONNECTION.endsWith("Connection")).toBe(true);
    expect(CONNECTION).not.toBe("Connection");
    expect(COMMAND.endsWith("Command")).toBe(true);
    expect(COMMAND).not.toBe("Command");

    const cmd = defs.find((f) => f.name === COMMAND)!;
    expect(cmd.dimensions).toEqual(["instance", "verb", "result"]);
  });

  it("shapes the connection family as Total/Interval counter pairs plus the gauge and duration sum", () => {
    const conn = familyDefs().find((f) => f.name === CONNECTION)!;
    const names = conn.measures.map((m) => m.name);
    for (const base of ["connectAttempts", "connectFailures", "reconnectAttempts", "connectionDrops"]) {
      expect(names).toContain(`${base}Total`);
      expect(names).toContain(`${base}Interval`);
    }
    expect(names).toContain("connectionState"); // the state gauge
    expect(names).toContain("connectedDurationMs"); // the connected-duration sum
  });

  it("resets interval counters on drain but not totals", () => {
    const p = new Pair();
    p.add(3);
    const out: Record<string, number> = {};
    p.drainInto(out, "x");
    expect(out.xTotal).toBe(3);
    expect(out.xInterval).toBe(3);

    p.add(2);
    const out2: Record<string, number> = {};
    p.drainInto(out2, "x");
    expect(out2.xTotal).toBe(5); // total is monotonic across emits
    expect(out2.xInterval).toBe(2); // interval resets to only what accrued since the last emit
  });
});

describe("DeviceMetrics emission", () => {
  it("counts only signals past the staleSignalSecs threshold", async () => {
    const rec = new RecordingMetrics();
    const health = new Health();
    const dm = new DeviceMetrics(rec, testConfig(), "plc-1", health, 30);
    const now = Date.now();
    dm.onSignalUpdate("fresh", now);
    dm.onSignalUpdate("stale", now - 120_000);
    await dm.emitPeriodic();
    expect(rec.last(HEALTH).staleSignals).toBe(1);
  });

  it("drains the health interval counters on emit", async () => {
    const rec = new RecordingMetrics();
    const health = new Health();
    const dm = new DeviceMetrics(rec, testConfig(), "plc-1", health, 30);
    health.readErrors = 3;
    health.reconnects = 2;
    await dm.emitPeriodic();
    const h = rec.last(HEALTH);
    expect(h.readErrors).toBe(3);
    expect(h.reconnects).toBe(2);
    // Reset after the emit so the next interval starts clean.
    expect(health.readErrors).toBe(0);
    expect(health.reconnects).toBe(0);
  });

  it("records a command into the Command family's (verb, result) counters", async () => {
    const rec = new RecordingMetrics();
    const dm = new DeviceMetrics(rec, testConfig(), "plc-1", new Health(), 30);
    dm.recordCommand("sb/status", true, 5);
    await dm.emitPeriodic();
    // Exactly one (verb, result) combo saw a request; the rest drained zero.
    const withOneRequest = rec.emissions.filter(
      (e) => e.name === COMMAND && e.values.commandRequestsTotal === 1,
    );
    expect(withOneRequest).toHaveLength(1);
    expect(withOneRequest[0].values.commandErrorsTotal).toBe(0);
  });
});

File: templates/java-protocol-adapter/src/test/java/com/mbreissi/testcomponent/MetricsTest.java
package <<PACKAGE>>;

import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Operational metrics parity: {@code southbound_health} is <b>exactly</b> the SOUTHBOUND.md §5 set, and
 * the two worked operational families follow the Total/Interval counter-pair pattern with
 * low-cardinality dimensions only.
 */
class MetricsTest {

    @Test
    void southboundHealthEmitsExactlyTheSection5MeasureSet() {
        // A second, independent copy of §5 — NOT the module const, so a wrong edit to one is caught.
        Set<String> section5 = new TreeSet<>(List.of(
                "connectionState", "publishLatencyMs", "pollLatencyMs", "readErrors", "staleSignals",
                "reconnects"));

        Metrics.FamilyDef health = Metrics.familyDefs().stream()
                .filter(f -> f.name().equals(Metrics.HEALTH)).findFirst().orElseThrow();
        Set<String> emitted = health.measures().stream().map(Metrics.MeasureDef::name)
                .collect(Collectors.toCollection(TreeSet::new));
        assertEquals(section5, emitted,
                "southbound_health must be the exact §5 set — no more, no less");

        Set<String> advertised = new TreeSet<>(List.of(Metrics.HEALTH_MEASURES));
        assertEquals(section5, advertised, "HEALTH_MEASURES must equal the §5 set");
    }

    @Test
    void operationalFamiliesAreNamedFromTheComponentAndLowCardinality() {
        List<Metrics.FamilyDef> defs = Metrics.familyDefs();
        List<String> names = defs.stream().map(Metrics.FamilyDef::name).toList();
        assertTrue(names.contains(Metrics.CONNECTION), "the Connection family is present");
        assertTrue(names.contains(Metrics.COMMAND), "the Command family is present");
        // Named from the component token — a fleet view separates adapters by name.
        assertTrue(Metrics.CONNECTION.endsWith("Connection") && !Metrics.CONNECTION.equals("Connection"));
        assertTrue(Metrics.COMMAND.endsWith("Command") && !Metrics.COMMAND.equals("Command"));

        Metrics.FamilyDef cmd = defs.stream().filter(f -> f.name().equals(Metrics.COMMAND))
                .findFirst().orElseThrow();
        assertEquals(List.of("instance", "verb", "result"), cmd.dimensions(),
                "closed, low-cardinality dims only");
    }

    @Test
    void theConnectionFamilyIsTheCounterPairPattern() {
        Metrics.FamilyDef conn = Metrics.familyDefs().stream()
                .filter(f -> f.name().equals(Metrics.CONNECTION)).findFirst().orElseThrow();
        List<String> names = conn.measures().stream().map(Metrics.MeasureDef::name).toList();
        for (String base : List.of("connectAttempts", "connectFailures", "reconnectAttempts",
                "connectionDrops")) {
            assertTrue(names.contains(base + "Total"), base + "Total present");
            assertTrue(names.contains(base + "Interval"), base + "Interval present");
        }
        assertTrue(names.contains("connectionState"), "the state gauge");
        assertTrue(names.contains("connectedDurationMs"), "the connected-duration sum");
    }

    @Test
    void intervalCountersResetOnDrainButTotalsDoNot() {
        Metrics.Pair p = new Metrics.Pair();
        p.add(3.0);
        Map<String, Float> out = new HashMap<>();
        p.drainInto(out, "x");
        assertEquals(3.0f, out.get("xTotal"));
        assertEquals(3.0f, out.get("xInterval"));

        p.add(2.0);
        Map<String, Float> out2 = new HashMap<>();
        p.drainInto(out2, "x");
        assertEquals(5.0f, out2.get("xTotal"), "total is monotonic across emits");
        assertEquals(2.0f, out2.get("xInterval"),
                "interval resets to only what accrued since the last emit");
    }

    @Test
    void staleSignalsCountsOnlySignalsPastTheThreshold() {
        long now = System.nanoTime();
        long thirtySecs = 30L * 1_000_000_000L;
        List<Long> lastUpdate = List.of(now, now - 120L * 1_000_000_000L);
        assertEquals(1L, Metrics.countStale(lastUpdate, now, thirtySecs),
                "only the signal older than staleSignalSecs is stale");
    }

    @Test
    void everyCommandVerbHasAMeasureBucketAndThereAreExactlyNine() {
        assertEquals(9, Metrics.COMMAND_VERBS.length);
        Set<String> verbs = new TreeSet<>(List.of(Metrics.COMMAND_VERBS));
        assertEquals(new TreeSet<>(List.of("sb/status", "sb/read", "sb/write", "sb/signals",
                "sb/browse", "sb/pause", "sb/resume", "reconnect", "repoll")), verbs);
    }
}
